Fetch and populate partition and domain with `shibaken`

jpculp opened the pull request:
**Description:**

Fetches the partition and domain from IMDS meta-data using the `shibaken` settings generator.

**Testing done:**

- [x] Launch host in `aws`.
- [x] Launch host in `aws-us-gov`.
- [x] Launch host in `aws-cn.
- [ ] Migrate from existing host.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
migrations: support 'aws.partition' setting for AWS variants

jpculp committed:
shibaken: fetch and return the AWS partition

In addition to returning a boolean of if a host is in a given partition,
`shibaken` can also return a string of the partition.

jpculp committed:
models: add 'aws.partition' setting

jpculp commented:
Moved logic from `early-boot-config` to `shibaken`.

stmcginnis commented on Release.toml:
```diff
@@ -1,4 +1,4 @@
-version = "1.15.0"
+version = "1.14.1"
```
We could either do this, or modify the cherry-pick back to `1.14.x`. I'll make a note to follow up to bump this back to 1.15.0 if we do this here. Only advantage is a cleaner backport.

stmcginnis commented on sources/api/migration/migrations/v1.14.1/aws-partition-setting/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+    if cfg!(variant_platform = "aws") {
+        migrate(AddSettingsMigration(&["settings.aws.partition"]))
+    } else {
+        // Non-AWS variants did not have any AWS setting until this point,
```
That's not actually true anymore. Since v.1.11.0 we've needed `settings.aws` to support AWS IAM Roles Anywhere.

https://github.com/bottlerocket-os/bottlerocket/pull/2542/commits/cedda93d79c849948c063dccedc076d1a05580fe

jpculp replied:
Ah, in that case I'll replace the else with a `NoOpMigration`.

stmcginnis commented on sources/api/migration/migrations/v1.14.1/shibaken-aws-partition/src/main.rs:
```diff
@@ -0,0 +1,26 @@
+            metadata: &["setting-generator"],
+        }]))
+    } else {
+        migrate(NoOpMigration)
```
This should be fine for non-AWS variants though, since there is no concept of a partition when it comes to IAM Roles Anywhere.

jpculp replied:
Like the region behavior in `early-boot-config`, partition should fallback to `aws` if IMDS fails to pull one.

jpculp commented:
- Reverted `Release.toml` version back to `1.15.0`.
- Replaced else `AddPrefixesMigration(vec!["settings.aws"])` with `NoOpMigration`.

bcressey commented on sources/api/migration/migrations/v1.14.1/aws-partition-setting/src/main.rs:
```diff
@@ -0,0 +1,22 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added a new setting for AWS partition.
+fn run() -> Result<()> {
+    if cfg!(variant_platform = "aws") {
+        migrate(AddSettingsMigration(&["settings.aws.partition"]))
+    } else {
+        migrate(NoOpMigration)
+    }
+}
```
This is not correct since `settings.aws.partition` will exist in all variants since `AwsSettings` is not conditionally compiled.
```suggestion
/// We added a new setting for AWS partition.
fn run() -> Result<()> {
    migrate(AddSettingsMigration(&["settings.aws.partition"]))
}
```

bcressey commented on sources/api/shibaken/README.md:
```diff
@@ -11,6 +11,7 @@ shibaken is used to fetch data from the instance metadata service (IMDS) in AWS.
 shibaken can:
 * Fetch and populate the admin container's user-data with authorized ssh keys from the IMDS.
 * Perform boolean queries about the AWS partition in which the host is located.
+* Fetch and populate the 'aws.partition' setting.
```
IMO we should go ahead and populate `aws.domain` as well while we're at it.

zmrow reviewed (approved):
Provided @bcressey 's feedback is addressed, this looks good.

jpculp committed:
imdsclient: add fetch_domain function

Adds another helper function to fetch the domain in which the
Bottlerocket instance resides.

jpculp committed:
shibaken: fetch and return the AWS domain

jpculp committed:
models: add 'aws.domain' setting

jpculp committed:
migrations: support 'aws.domain' setting for AWS variants

jpculp commented:
- Fixed conditionally compiled migration.
- Added commits to support a domain setting.

jpculp commented:
I think I might need to tweak this a bit. I'll open a new PR when I do.
